src/main/java/jenkins/branch/MultiBranchProject.java

```diff
@@ -2137,8 +2137,9 @@ private String getProjectDisplayName(@NonNull P project, @NonNull String rawName
                     .orElse(null);
             }
 
+            // Default to displaying the project's display name if a trait hasn't been provided
             if (naming == null) {
-                return rawName;
+                naming = MultiBranchProjectDisplayNamingStrategy.RAW_AND_OBJECT_DISPLAY_NAME;
```

Comment:
This is the line causing the failure, previously it returned but now on L2146 it retrieves the `ObjectMetadataAction` which needs to load more information which is failing.

The `ClassRule` will be needed for performance as it only loads one instance of Jenkins for the whole test. 1s per test rather than like 10s.

I haven't figured out what's causing this though.

@jtnord any ideas?

Reply:
a very quick rough guess would be that there have been no builds of this project yet and the project may not actually be persisted at this point?  (or the metadata is attached after the displayname has used
This now then falls into a lookup,  as the namings strategy requires the "displayName" which means the metadata needs to be looked up, which during creation is not present.

```
java.lang.AssertionError
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:644)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.getByNumber(AbstractLazyLoadRunMap.java:543)
	at jenkins.model.lazy.LazyBuildMixIn.getBuildByNumber(LazyBuildMixIn.java:240)
	at hudson.model.AbstractProject.getBuildByNumber(AbstractProject.java:931)
	at hudson.model.AbstractProject.getBuildByNumber(AbstractProject.java:137)
	at jenkins.model.PeepholePermalink.resolve(PeepholePermalink.java:107)
	at hudson.model.Job.getLastSuccessfulBuild(Job.java:945)
	at jenkins.tasks.SimpleBuildStep$LastBuildActionFactory.createFor(SimpleBuildStep.java:198)
	at jenkins.tasks.SimpleBuildStep$LastBuildActionFactory.createFor(SimpleBuildStep.java:184)
	at hudson.model.Actionable.createFor(Actionable.java:115)
	at hudson.model.Actionable.getAction(Actionable.java:332)
	at jenkins.branch.MultiBranchProject$SCMHeadObserverImpl.getProjectDisplayName(MultiBranchProject.java:2146)
	at jenkins.branch.MultiBranchProject$SCMHeadObserverImpl.observeNew(MultiBranchProject.java:2113)
	at jenkins.branch.MultiBranchProject$SCMHeadObserverImpl.observe(MultiBranchProject.java:2023)
	at jenkins.scm.api.trait.SCMSourceRequest.process(SCMSourceRequest.java:357)
	at jenkins.scm.api.trait.SCMSourceRequest.process(SCMSourceRequest.java:286)
	at jenkins.scm.impl.mock.MockSCMSource.retrieve(MockSCMSource.java:137)
	at jenkins.scm.api.SCMSource._retrieve(SCMSource.java:372)
	at jenkins.scm.api.SCMSource.fetch(SCMSource.java:282)
	at jenkins.branch.MultiBranchProject.computeChildren(MultiBranchProject.java:654)
	at com.cloudbees.hudson.plugins.folder.computed.ComputedFolder.updateChildren(ComputedFolder.java:265)
	at com.cloudbees.hudson.plugins.folder.computed.FolderComputation.run(FolderComputation.java:167)
	at jenkins.branch.MultiBranchProject$BranchIndexing.run(MultiBranchProject.java:1057)
```

IIRC `updateChildren" also creates the new Project. 

NB: there are also some traits where a job gets created for a PR/Branch but will prevent the initial build from triggering (so this could also impact if a build is needed to retreive the info)

Reply:
That's what I was wondering but I've tried what Jan said and switching it from a `ClassRule` to a `Rule` does fix the tests. (as does running tests individually).

I've also tried adding sleeps in some places to see if something is just slow but it didn't help.

Any other thoughts?

> IIRC `updateChildren" also creates the new Project.

is that a suggestion for an alternative method? if so its package-private.

Reply:
> is that a suggestion for an alternative method? if so its package-private.

no - its at the base of the stack - so was just that it may be during creating projects and not just updating existing ones.

Reply:
I think when you build a multi branch project thats just running indexing. So its after creation and its the the `scheduleBuild2` call that is triggering this.